src/CADL.Extension/Emitter.Csharp/src/lib/operation.ts

```diff
@@ -0,0 +1,446 @@
+                    models,
+                    enums
+                );
+            } else {
+                const cadlType = getEffectiveSchemaType(program, body.type);
+                const inputType: InputType = getInputType(
+                    program,
+                    cadlType,
+                    models,
+                    enums
+                );
+                type = inputType;
+            }
+        }
+
+        const headers = response.responses[0]?.headers;
+        const responseHeaders: HttpResponseHeader[] = [];
+        if (headers) {
+            for (const key of Object.keys(headers)) {
+                responseHeaders.push({
+                    Name: key,
+                    NameInResponse: headers[key].name,
+                    Description: getDoc(program, headers[key]) ?? "",
+                    Type: getInputType(
+                        program,
+                        headers[key].type,
+                        models,
+                        enums
+                    )
+                } as HttpResponseHeader);
+            }
+        }
+
+        return {
+            StatusCodes: status,
+            BodyType: type,
+            BodyMediaType: BodyMediaType.Json,
+            Headers: responseHeaders,
+            IsErrorResponse: isErrorModel(program, response.type)
+        } as OperationResponse;
+    }
+
+    function loadOperationLongRunning(
+        program: Program,
+        op: HttpOperation,
+        resourceOperation?: ResourceOperation
+    ): OperationLongRunning | undefined {
+        if (!isLroOperation(program, op.operation)) return undefined;
+
+        const finalResponse = loadLongRunningFinalResponse(
+            program,
+            op,
+            resourceOperation
+        );
+        if (finalResponse === undefined) return undefined;
+
+        return {
+            FinalStateVia: OperationFinalStateVia.Location, // data plane only supports `location`
+            FinalResponse: finalResponse
+        } as OperationLongRunning;
+    }
+
+    function loadLongRunningFinalResponse(
+        program: Program,
+        op: HttpOperation,
+        resourceOperation?: ResourceOperation
+    ): OperationResponse | undefined {
+        let finalResponse: any | undefined;
+        for (const response of op.responses) {
+            if (response.statusCode === "200") {
+                finalResponse = response;
+                break;
+            }
+            if (response.statusCode === "204") {
+                finalResponse = response;
+            }
+        }
+
+        if (finalResponse !== undefined) {
+            return loadOperationResponse(
+                program,
+                finalResponse,
+                resourceOperation
+            );
+        }
+
+        return loadOperationResponse(
+            program,
+            op.responses[0],
+            resourceOperation
+        );
+    }
+
+    function isLroOperation(program: Program, op: Operation) {
```

Comment:
since `Lro` stands for long running operation, we might need to rename this to either `isLro` or `isLongRunningOperation`